File: backend/app/controllers/auth_controller.py
from flask import Blueprint, request, jsonify, make_response, session
from services.auth_service import authenticate_google_user
from config import Config
from services.user_service import login_user, register_user
import logging
from services.user_service import generate_signed_session_id, redis_client

auth_bp = Blueprint("auth", __name__)

@auth_bp.route("/auth/google/callback", methods=["POST"])
def google_callback():
    data = request.get_json()
    id_token_from_google = data.get("id_token")
    if not id_token_from_google:
        return jsonify({"error": "缺少 id_token"}), 400

    user, error = authenticate_google_user(id_token_from_google)
    if error:
        return jsonify({"error": f"Google OAuth 處理失敗: {error}"}), 400

    response = make_response(jsonify({
        "message": "Google 登入成功",
        "username": user.username,
        "role": user.role,
        "last_login": user.last_login.isoformat() if user.last_login else None,
        "login_count": user.login_count
    }))
    return response

@auth_bp.route("/register", methods=["POST"])
def register():
    data = request.json
    register_response = register_user(data)

    if "error" in register_response:
        return jsonify({"error": register_response["error"]}), 400

    response = make_response(jsonify({"message": "註冊成功", "user": register_response["user"].username, "role": "user"}))
    response.set_cookie("user_id", str(register_response["user"].id), httponly=True, secure=True, samesite="Strict")
    response.set_cookie("role", "user", httponly=True, secure=True, samesite="Strict")
    return response, 201

@auth_bp.route("/login", methods=["POST"])
def login():
    data = request.json
    login_response = login_user(data)

    if "error" in login_response:
        return jsonify({"error": login_response["error"]}), 400

    response = make_response(jsonify({
        "message": "登入成功",
        "role": login_response["role"],
        "last_login": login_response["last_login"].isoformat() if login_response["last_login"] else None,
        "login_count": login_response["login_count"]
    }))
    return response

@auth_bp.route("/logout")
def logout():
    response = make_response(jsonify({"message": "登出成功"}))
    response.set_cookie("user_id", "", expires=0)
    response.set_cookie("role", "", expires=0)
    return response

File: backend/golang-app/controllers/auth_controller.go
package controllers

import (
	"fmt"
	"golang-app/models"
	"golang-app/services"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

func RegisterAuthRoutes(r *gin.Engine) {
	auth := r.Group("/auth")
	{
		auth.POST("/google/callback", GoogleCallback)
		auth.POST("/register", Register)
		auth.POST("/login", Login)
	}
	r.POST("/login", Login)
	r.POST("/register", Register)
	r.POST("/google-callback", GoogleCallback)
}

func GoogleCallback(c *gin.Context) {
	var data map[string]interface{}
	if err := c.ShouldBindJSON(&data); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid JSON"})
		return
	}

	idTokenFromGoogle, exists := data["id_token"].(string)
	if !exists || idTokenFromGoogle == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "缺少 id_token"})
		return
	}

	user, err := services.AuthenticateGoogleUser(idTokenFromGoogle, models.DB)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("Google OAuth 處理失敗: %v", err)})
		return
	}

	response := gin.H{
		"message":     "Google 登入成功",
		"username":    user.Username,
		"role":        user.Role,
		"last_login":  user.LastLogin,
		"login_count": user.LoginCount,
	}

	c.SetCookie("user_id", fmt.Sprintf("%d", user.ID), 3600, "/", "", false, true)
	c.SetCookie("role", string(user.Role), 3600, "/", "", false, true)

	c.JSON(http.StatusOK, response)
}

func Register(c *gin.Context) {
	var registerRequest services.RegisterRequest

	if err := c.ShouldBindJSON(&registerRequest); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "請提供註冊資料"})
		return
	}

	registerResponse, err := services.RegisterUser(registerRequest, models.DB)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("註冊失敗: %v", err)})
		return
	}

	userObj, ok := registerResponse["user"].(*models.User)
	if !ok {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "用戶資料轉換失敗"})
		return
	}

	c.SetCookie("user_id", fmt.Sprintf("%d", userObj.ID), 3600, "/", "", false, true)
	c.SetCookie("role", "user", 3600, "/", "", false, true)

	c.JSON(http.StatusOK, gin.H{
		"message": "註冊成功",
		"user":    userObj.Username,
		"role":    "user",
	})
}

func Login(c *gin.Context) {
	var loginRequest services.LoginRequest

	if err := c.ShouldBindJSON(&loginRequest); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "請提供登入資料"})
		return
	}

	loginResponse, err := services.LoginUser(loginRequest, models.DB)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("登入失敗: %v", err)})
		return
	}

	userID := fmt.Sprintf("%v", loginResponse["user_id"])
	role, _ := loginResponse["role"].(string)
	lastLogin, _ := loginResponse["last_login"].(time.Time)
	loginCount, _ := loginResponse["login_count"].(int)

	c.SetCookie("user_id", userID, 3600, "/", "", false, true)
	c.SetCookie("role", role, 3600, "/", "", false, true)

	c.JSON(http.StatusOK, gin.H{
		"message":     "登入成功",
		"role":        role,
		"last_login":  lastLogin,
		"login_count": loginCount,
	})
}

func Logout(c *gin.Context) {
	c.SetCookie("user_id", "", -1, "/", "", false, true)
	c.SetCookie("role", "", -1, "/", "", false, true)

	c.JSON(http.StatusOK, gin.H{
		"message": "登出成功",
	})
}
